craft_cli/completion/completion.py

```diff
@@ -193,14 +192,16 @@ def all_args(self) -> str:
         return " ".join([a.flag_list for a in self.args])
 
 
-def complete(shell_cmd: str, get_dispatcher: Callable[[], craft_cli.Dispatcher]) -> str:
+def complete(
+    shell_cmd: str, get_app_info: Callable[[], Tuple[craft_cli.Dispatcher, Dict[str, Any]]]
```

Comment:
I think the second element of the tuple should be `Dict[str, Any] | None`, following what BaseCommand [accepts](https://github.com/canonical/craft-cli/blob/main/craft_cli/dispatcher.py#L158).

Reply:
since you use this Tuple type in a lot of places it might make sense to create something like a `DispatcherAndConfig` type